File: typescript/packages/mechanisms/evm/src/batch-settlement/client/eip3009.ts
import { PaymentRequirements, PaymentPayloadResult } from "@x402/core/types";
import { getAddress } from "viem";
import { ClientEvmSigner } from "../../signer";
import { ChannelConfig, BatchSettlementDepositPayload } from "../types";
import { ERC3009_DEPOSIT_COLLECTOR_ADDRESS, receiveAuthorizationTypes } from "../constants";
import { createNonce, getEvmChainId } from "../../utils";
import { signVoucher } from "./voucher";
import { computeChannelId } from "../utils";
import { buildErc3009DepositNonce } from "../encoding";

/**
 * Creates a deposit payload that bundles an ERC-3009 `receiveWithAuthorization` approval
 * together with a cumulative voucher signature.
 *
 * When the facilitator submits this payload onchain, the contract atomically transfers
 * tokens from the payer into the channel and records the initial voucher.
 *
 * @param signer - Client wallet used to sign the ERC-3009 authorization (`from` = payer).
 * @param x402Version - Protocol version to embed in the payload envelope.
 * @param paymentRequirements - Server-provided payment requirements (asset, network, amount, etc.).
 * @param channelConfig - Immutable channel configuration (payer, receiver, token, …).
 * @param depositAmount - Number of tokens (decimal string) to deposit into the channel.
 * @param maxClaimableAmount - Cumulative ceiling for the accompanying voucher.
 * @param voucherSigner - Optional key that signs the voucher; defaults to `signer` (same as payer).
 * @returns A {@link PaymentPayloadResult} containing the signed deposit + voucher payload.
 */
export async function createBatchSettlementEIP3009DepositPayload(
  signer: ClientEvmSigner,
  x402Version: number,
  paymentRequirements: PaymentRequirements,
  channelConfig: ChannelConfig,
  depositAmount: string,
  maxClaimableAmount: string,
  voucherSigner?: ClientEvmSigner,
): Promise<PaymentPayloadResult> {
  const salt = createNonce();
  const now = Math.floor(Date.now() / 1000);
  const chainId = getEvmChainId(paymentRequirements.network);

  if (!paymentRequirements.extra?.name || !paymentRequirements.extra?.version) {
    throw new Error(
      `EIP-712 domain parameters (name, version) are required in payment requirements for asset ${paymentRequirements.asset}`,
    );
  }

  const { name, version } = paymentRequirements.extra;

  const channelId = computeChannelId(channelConfig, paymentRequirements.network);

  const erc3009Nonce = buildErc3009DepositNonce(channelId, salt);

  const signature = await signer.signTypedData({
    domain: {
      name,
      version,
      chainId,
      verifyingContract: getAddress(paymentRequirements.asset),
    },
    types: receiveAuthorizationTypes,
    primaryType: "ReceiveWithAuthorization",
    message: {
      from: getAddress(signer.address),
      to: getAddress(ERC3009_DEPOSIT_COLLECTOR_ADDRESS),
      value: BigInt(depositAmount),
      validAfter: BigInt(now - 600),
      validBefore: BigInt(now + paymentRequirements.maxTimeoutSeconds),
      nonce: erc3009Nonce,
    },
  });

  const vSigner = voucherSigner ?? signer;
  const voucher = await signVoucher(
    vSigner,
    channelId,
    maxClaimableAmount,
    paymentRequirements.network,
  );

  const payload: BatchSettlementDepositPayload = {
    type: "deposit",
    channelConfig,
    voucher,
    deposit: {
      amount: depositAmount,
      authorization: {
        erc3009Authorization: {
          validAfter: (now - 600).toString(),
          validBefore: (now + paymentRequirements.maxTimeoutSeconds).toString(),
          salt,
          signature,
        },
      },
    },
  };

  return {
    x402Version,
    payload,
  };
}

File: python/x402/mechanisms/evm/batch_settlement/client/eip3009.py
"""ERC-3009 deposit payload construction (client side)."""

from __future__ import annotations

import time

try:
    from eth_utils import to_checksum_address
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .....schemas import PaymentRequirements
from ....evm.signer import ClientEvmSigner
from ...utils import create_nonce, get_evm_chain_id
from ..constants import ERC3009_DEPOSIT_COLLECTOR_ADDRESS, RECEIVE_AUTHORIZATION_TYPES
from ..encoding import build_erc3009_deposit_nonce
from ..types import (
    ChannelConfig,
    DepositAuthorization,
    DepositFields,
    DepositPayload,
    Erc3009Authorization,
)
from ..utils import coerce_bytes32, compute_channel_id
from .voucher import sign_voucher


def create_batch_settlement_eip3009_deposit_payload(
    signer: ClientEvmSigner,
    requirements: PaymentRequirements,
    channel_config: ChannelConfig,
    deposit_amount: str | int,
    max_claimable_amount: str | int,
    voucher_signer: ClientEvmSigner | None = None,
) -> DepositPayload:
    """Create a deposit payload bundling an ERC-3009 authorization + voucher."""
    extra = requirements.extra or {}
    name = extra.get("name")
    version = extra.get("version")
    if not name or not version:
        raise ValueError(
            "EIP-712 domain parameters (name, version) are required in payment "
            f"requirements for asset {requirements.asset}"
        )

    salt = create_nonce()
    now = int(time.time())
    valid_after = now - 600
    valid_before = now + int(requirements.max_timeout_seconds)
    chain_id = get_evm_chain_id(str(requirements.network))
    channel_id = compute_channel_id(channel_config, str(requirements.network))

    erc3009_nonce = build_erc3009_deposit_nonce(channel_id, salt)

    domain = {
        "name": name,
        "version": version,
        "chainId": chain_id,
        "verifyingContract": to_checksum_address(requirements.asset),
    }
    message = {
        "from": to_checksum_address(signer.address),
        "to": to_checksum_address(ERC3009_DEPOSIT_COLLECTOR_ADDRESS),
        "value": int(deposit_amount),
        "validAfter": valid_after,
        "validBefore": valid_before,
        "nonce": coerce_bytes32(erc3009_nonce),
    }
    sig_bytes = signer.sign_typed_data(
        domain, RECEIVE_AUTHORIZATION_TYPES, "ReceiveWithAuthorization", message
    )
    signature = "0x" + sig_bytes.hex() if not sig_bytes.hex().startswith("0x") else sig_bytes.hex()

    voucher = sign_voucher(
        voucher_signer or signer,
        channel_id,
        max_claimable_amount,
        str(requirements.network),
    )

    payload = DepositPayload()
    payload.channel_config = channel_config
    payload.voucher = voucher
    payload.deposit = DepositFields(
        amount=str(int(deposit_amount)),
        authorization=DepositAuthorization(
            erc3009_authorization=Erc3009Authorization(
                valid_after=str(valid_after),
                valid_before=str(valid_before),
                salt=salt,
                signature=signature,
            ),
        ),
    )
    return payload


__all__ = ["create_batch_settlement_eip3009_deposit_payload"]
